src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            if (parameterWithMatchingName.Ordinal >= arguments.Count)
+            {
+                return null;
+            }
+
+            // Multiple arguments could have been converted to a single params array, 
+            // so there wouldn't be a corresponding argument
+            if (parameterWithMatchingName.IsParams && parameters.Length != arguments.Count)
+            {
+                return null;
+            }
+
+            return arguments[parameterWithMatchingName.Ordinal];
+        }
+
+        private IParameterSymbol GetParameterWithMatchingName(ImmutableArray<IParameterSymbol> parameters, string searchArgumentName)
+        {
+            foreach (var p in parameters)
+            {
+                if (p.Name.Equals(searchArgumentName))
+                {
+                    return p;
+                }
+            }
+
+            return null;
+        }
+
+        protected SyntaxNode TryGetFormatStringLiteralExpressionSyntax(
+            SemanticModel semanticModel,
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            var formatArgumentSyntax = TryGetArgument(
+                semanticModel,
+                NameOfFormatStringParameter,
+                arguments,
+                parameters);
+
+            if (formatArgumentSyntax == null)
+            {
+                return null;
+            }
+
+            if (!ArgumentExpressionIsStringLiteral(formatArgumentSyntax))
+            {
+                return null;
+            }
+
+            return GetArgumentExpression(formatArgumentSyntax);
+        }
+
+        protected static IMethodSymbol TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
+            SymbolInfo symbolInfo)
+        {
+
+            if (symbolInfo.Symbol == null)
+            {
+                return null;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return null;
+            }
+
+            var containingType = (INamedTypeSymbol)symbolInfo.Symbol.ContainingSymbol;
+
+            if (containingType.SpecialType != SpecialType.System_String)
+            {
+                return null;
+            }
+
+            var method = (IMethodSymbol)symbolInfo.Symbol;
+
+            if (method.Arity > 0)
+            {
+                return null;
+            }
+
+            return method;
+        }
+
+        private bool FormatCallWorksAtRuntime(string formatString, int numberOfPlaceholderArguments)
+        {
+            var testArray = new object[numberOfPlaceholderArguments];
+            for (var i = 0; i < numberOfPlaceholderArguments; i++)
+            {
+                testArray[i] = "test";
+            }
+
+            try
+            {
+                string.Format(formatString, testArray);
+            }
+            catch
```

Comment:
If it wasn't that we regularly complain about exceptions being thrown by other teams in "regular" paths I wouldn't care too much. But we do and this will potentially make debugging really annoying.

This also makes me realize that we're presuming here the implementation of string.Format in the VS runtime matches the implementation of your target framework, which may not be the case. More generally, that makes me wish we had a way to flow this analyzer with the framework (i.e. put it in roslyn-analyzers repo instead?) so any behavior updates could be handled appropriately.

There *is* one very subtle assumption being made here which is this won't get tripped up with any format specifiers, i.e. that string always takes any format specifier. If that's ever *not* the case then this is actually broken.

The performance is also icky: it means in _correct_ code we're allocating all sorts of strings only to see if we could allocate such a string.

I can't say this approach is obviously better to me.